### ISSUE TITLE
State exception when executing procedure with runtime error

### ISSUE THREAD

**Issue by Hjorthen** [mysql-net/MySqlConnector#423]
Hi. I've been using MySQLConnector for some time now. However, I've run into an exception `System.InvalidOperationException : 'Expected state to be Failed but was Connected.'` which fires on the last line of the following: 
```
 MySqlCommand command = conn.Get().CreateCommand();
 command.CommandText = "getUserRole";
 command.CommandType = CommandType.StoredProcedure;
 command.Parameters.AddWithValue("@userID", username);
 command.Parameters.AddWithValue("@projectID", -1);
 command.Parameters.AddWithValue("@userRole", MySqlDbType.Int32).Direction = ParameterDirection.Output;
 await command.ExecuteNonQueryAsync();
```
Stack trace:
```
at MySqlConnector.Core.ServerSession.VerifyState(State state) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 882
   at MySqlConnector.Core.ServerSession.FinishQuerying() in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 154
   at MySql.Data.MySqlClient.MySqlDataReader.DoClose() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 419
   at MySql.Data.MySqlClient.MySqlDataReader.Dispose(Boolean disposing) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 266
   at MySqlConnector.Core.TextCommandExecutor.<ExecuteNonQueryAsync>d__1.MoveNext() in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextCommandExecutor.cs:line 33
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
```
Any suggestions for what the cause might be or how to prevent it?

**Comment by bgrainger** [comment]
Are you able to copy the `CREATE PROCEDURE` statement here? That may help me reproduce the issue.

Which version of MySQL Server are you using?

**Comment by Hjorthen** [comment]
It happens whenever there is a runtime error in the procedure(such as duplicate key)
This is the latest procedure I have encountered it with(same error as previously stated):
```

**Comment by bgrainger** [comment]
FWIW there is a bug in your first code snippet:

```
command.Parameters.AddWithValue("@userRole", MySqlDbType.Int32).Direction = ParameterDirection.Output;
```

The second parameter to `AddWithValue` is used to initialize the `MySqlParameter.Value` property, not the `MySqlParameter.MySqlDbType` property. In this case it doesn't really matter because this is an `out` parameter and its value will get overwritten anyway. However, it looks like you might be trying to set `.MySqlDbType` and that will not happen. (AFAICT this is the same in MySqlConnector and Connector/NET.)

**Comment by bgrainger** [comment]
You probably wanted `.Add("@userRole", MySqlDbType.Int32)` but that overload doesn't exist yet; created #424 to add it.

**Comment by Hjorthen** [comment]
Right. What about raising the exception on C# level?

**Comment by bgrainger** [comment]
Never mind, I just saw that you posted above:

> (Tried 0.34.0, no change)

In that case, a self-contained repro would be greatly appreciated.

**Comment by Hjorthen** [comment]
I don't even know where to start. What could cause the session to become in an invalid state?

**Comment by bgrainger** [comment]
Once I have a repro, I can figure that out. I'm just not able to write code that reproduces the exception you're reporting.

**Comment by bgrainger** [comment]
Can you show me the definition of the `getUserRole` stored procedure?

**Comment by Hjorthen** [comment]
I changed `getUserRole` to a function instead of a procedure. The runtime error for it was that it was trying to retrieve a column that didn't exist(typo). It seems to be an issue with all runtime errors. 

The code I use for calling create project is here: 
```

        public async Task<int> CreateProject(string name, string userId, string comment)
        {
            using(Connection conn = await connection.Create())
            {
                MySqlCommand command = conn.Get().CreateCommand();
                command.CommandText = "createProject";
                command.CommandType = CommandType.StoredProcedure;
                command.Parameters.AddWithValue("@nameparam", name);
                command.Parameters.AddWithValue("@userid", userId);
                command.Parameters.AddWithValue("@commentparam", comment);
                command.Parameters.Add("@id", DbType.Int32).Direction = ParameterDirection.Output;
                await command.ExecuteNonQueryAsync();
                return (int)command.Parameters["@id"].Value;
            }
        }
```

**Comment by bgrainger** [comment]
I believe I've identified the cause of this bug.

**Comment by Hjorthen** [comment]
Something I can do to prevent it?

**Comment by bgrainger** [comment]
AFAICT, this happens when executing the stored procedure or function causes an exception (which is then masked by the exception you reported in the original bug report). So the only workaround I can think of is: "don't cause that exception in the first place". To find out what it is, you'd need to debug with "Break on all first-chance exceptions", then look at it in the debugger.

A real fix should be shipping soon.

**Comment by Hjorthen** [comment]
I'll just make the procedures return a "magic" number for now. Thanks for your support!

**Comment by bgrainger** [comment]
Fixed in [0.34.1](https://github.com/mysql-net/MySqlConnector/releases/tag/0.34.1).
